File: apps/ingestion/src/norvii_ingestion/publication/persistence/postgres.py
"""Canonical PostgreSQL connectivity adapter."""

from __future__ import annotations

from typing import TYPE_CHECKING, ClassVar

import psycopg

from norvii_ingestion.publication.persistence.errors import PersistenceConnectionError

if TYPE_CHECKING:
    from norvii_ingestion.publication.persistence.config import PostgresConfiguration


class PostgresStore:
    """Verify canonical storage through the production psycopg driver."""

    name: ClassVar[str] = "PostgreSQL"

    def __init__(self, connection: psycopg.Connection[tuple[object, ...]]) -> None:
        self._connection = connection

    @classmethod
    def connect(cls, configuration: PostgresConfiguration, timeout_seconds: int) -> PostgresStore:
        """Open an authenticated connection with bounded connect and statement timeouts."""
        try:
            connection = psycopg.connect(
                host=configuration.host,
                port=configuration.port,
                dbname=configuration.database,
                user=configuration.user,
                password=configuration.password,
                connect_timeout=timeout_seconds,
                options=f"-c statement_timeout={timeout_seconds * 1000}",
                autocommit=True,
            )
        except psycopg.Error as error:
            raise PersistenceConnectionError("Connect to PostgreSQL failed.") from error
        return cls(connection)

    def verify(self) -> None:
        """Execute a constant read-only SQL query."""
        try:
            with self._connection.cursor() as cursor:
                cursor.execute("SELECT 1")
                row = cursor.fetchone()
        except psycopg.Error as error:
            raise PersistenceConnectionError(
                "Execute PostgreSQL readiness query failed."
            ) from error
        if row != (1,):
            raise PersistenceConnectionError(
                "PostgreSQL readiness query returned an unexpected value."
            )

    def close(self) -> None:
        """Release the canonical-store connection."""
        self._connection.close()

File: apps/api/internal/platform/persistence/postgres.go
package persistence

import (
	"context"
	"fmt"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

// PostgresStore verifies the canonical store through the production pgx driver.
type PostgresStore struct {
	connection *pgx.Conn
}

// OpenPostgresPool opens the bounded caller-owned pool used by online repositories.
func OpenPostgresPool(ctx context.Context, config PostgresConfig) (*pgxpool.Pool, error) {
	connectionConfig, err := newPGXConfig(config)
	if err != nil {
		return nil, err
	}
	poolConfig, err := pgxpool.ParseConfig("sslmode=disable")
	if err != nil {
		return nil, fmt.Errorf("create PostgreSQL pool configuration: %w", err)
	}
	poolConfig.ConnConfig = connectionConfig
	pool, err := pgxpool.NewWithConfig(ctx, poolConfig)
	if err != nil {
		return nil, fmt.Errorf("connect PostgreSQL pool: %w", err)
	}
	if err := pool.Ping(ctx); err != nil {
		pool.Close()
		return nil, fmt.Errorf("verify PostgreSQL pool: %w", err)
	}
	return pool, nil
}

// OpenPostgresStore opens an authenticated canonical-store connection.
func OpenPostgresStore(ctx context.Context, config PostgresConfig) (*PostgresStore, error) {
	connectionConfig, err := newPGXConfig(config)
	if err != nil {
		return nil, err
	}
	connection, err := pgx.ConnectConfig(ctx, connectionConfig)
	if err != nil {
		return nil, fmt.Errorf("connect to PostgreSQL: %w", err)
	}
	return &PostgresStore{connection: connection}, nil
}

// Name returns the safe diagnostic identity for the canonical store.
func (store *PostgresStore) Name() string {
	return "PostgreSQL"
}

// Verify executes a constant read-only query.
func (store *PostgresStore) Verify(ctx context.Context) error {
	var ready int
	if err := store.connection.QueryRow(ctx, "SELECT 1").Scan(&ready); err != nil {
		return fmt.Errorf("execute canonical readiness query: %w", err)
	}
	if ready != 1 {
		return fmt.Errorf("canonical readiness query returned an unexpected value")
	}
	return nil
}

// Close releases the pgx connection.
func (store *PostgresStore) Close(ctx context.Context) error {
	return store.connection.Close(ctx)
}
